src/api/datalayouts.jl

```diff
@@ -0,0 +1,209 @@
+    @assert ndims(O) == ndims(array)
+    RRTMGPData{O, A, eltype(array), ndims(array)}(array)
+end
+
+"""
+    NVCData(gp::RRTMGPGridParams, ::Type{T}; N, nlay)
+    NVCData(gp::RRTMGPGridParams; N, nlay)
+
+Construct a `RRTMGPData` object with index order (N, vertical level, column).
+"""
+NVCData(gp::RRTMGPGridParams, ::Type{T}; N = 1, nlay) where {T} = RRTMGPData(NVCOrder, gp, T, N, nlay, gp.ncol)
+
+NVCData(gp::RRTMGPGridParams; N = 1, nlay) = NVCData(gp, eltype(gp); N, nlay)
+
+"""
+    VCData(gp::RRTMGPGridParams, ::Type{T}; nlay)
+    VCData(gp::RRTMGPGridParams; nlay)
+    VCData(fn, gp::RRTMGPGridParams; nlay)
+
+Construct a `RRTMGPData` object with index order (vertical level, column).
+"""
+VCData(gp::RRTMGPGridParams, ::Type{T}; nlay) where {T} = RRTMGPData(VCOrder, gp, T, nlay, gp.ncol)
+
+VCData(gp::RRTMGPGridParams; nlay) = VCData(gp, eltype(gp); nlay)
+
+"""
+    NCData(gp::RRTMGPGridParams, ::Type{T}; N)
+    NCData(gp::RRTMGPGridParams; N)
+
+Construct a `RRTMGPData` object with index order (N, column).
+"""
+NCData(gp::RRTMGPGridParams, ::Type{T}; N = 1) where {T} = RRTMGPData(NCOrder, gp, T, N, gp.ncol)
+NCData(gp::RRTMGPGridParams; N = 1) = NCData(gp, eltype(gp); N)
+
+"""
+    NData(gp::RRTMGPGridParams, ::Type{T}; N)
+    NData(gp::RRTMGPGridParams; N)
+
+Construct a `RRTMGPData` object with index order (N).
+"""
+NData(gp::RRTMGPGridParams, ::Type{T}; N = 1) where {T} = RRTMGPData(NOrder, gp, T, N)
+NData(gp::RRTMGPGridParams; N = 1) = NData(gp, eltype(gp); N)
+
+"""
+    RRTMGPData(
+    	::Type{<:AbstractIndexOrdering},
+    	grid_params::RRTMGPGridParams,
+    	::Type{Eltype},
+    	s...
+	)
+
+Construct a `RRTMGPData` object of type `O` and element type `Eltype`, with
+dimensions given by `s...`. The data is allocated using the
+`ClimaComms.array_type` appropriate for the grid.
+"""
+function RRTMGPData(::Type{O}, grid_params::RRTMGPGridParams, ::Type{Eltype}, s...) where {O, Eltype}
+    DA = ClimaComms.array_type(grid_params)
+    return RRTMGPData{O}(DA{Eltype}(undef, s...))
+end
+
+Base.parent(data::RRTMGPData) = data.array
+Base.size(data::RRTMGPData) = Base.size(parent(data))
+Base.length(data::RRTMGPData) = Base.length(parent(data))
+Base.eltype(data::RRTMGPData) = Base.eltype(parent(data))
+
+Base.Array(data::RRTMGPData{O}) where {O} = RRTMGPData{O}(Base.Array(parent(data)))
+Base.copyto!(data::RRTMGPData, array::AbstractArray) = Base.copyto!(parent(data), array)
+
+Base.@propagate_inbounds Base.getindex(data::RRTMGPData, I::CartesianIndex) = Base.getindex(parent(data), I.I...)
+
+Base.@propagate_inbounds Base.getindex(data::RRTMGPData{NVCOrder}, i, v, c) = Base.getindex(parent(data), i, v, c)
+Base.@propagate_inbounds Base.getindex(data::RRTMGPData{VCOrder}, v, c) = Base.getindex(parent(data), v, c)
+Base.@propagate_inbounds Base.getindex(data::RRTMGPData{NCOrder}, i, c) = Base.getindex(parent(data), i, c)
+Base.@propagate_inbounds Base.getindex(data::RRTMGPData{NOrder}, i) = Base.getindex(parent(data), i)
```

Comment:
```suggestion
Base.@propagate_inbounds Base.getindex(data::RRTMGPData{O}, indices...) where {O} =
    Base.getindex(parent(data), indices...)
```
I don't see how the verbosity here is helpful. Same for `setindex!`.

Reply:
But that edge case only happens with 1s, and any other index triggers a BoundaError. Is there a specific reason you're trying to prevent people from calling array[1, 1, ...]?

Reply:
True, I'm fine with simplifying it and relying on boundschecks.